# Comp790: Computational Biology

## Logistical Details
Instructor: Natalie Stanley

Email: stanleyn@stanford.edu

Time: Tuesday/Thursday 9:30am-10:45am

Office Hours: TBD

## Description
Moden, high-throughput assays allow us to efficiently profile multiple biological processes at a systems-level across a set of patient samples. As a result, these technologies generate an abundance of detailed information that needs to be extracted, analyzed and interpreted. In this course we will discuss the methodology used to analyze (process, engineer features, combine, etc.) data generated by some of the most cutting-edge technologies, such as proteomics, single-cell assays, and imaging in biomedicine.  We will further discuss the intersection of modern machine learning techniques and how they can be applied to most effectively extract information from these modalities and objectively benchmark existing algorithms for an improved understanding of human health and disease.

## Prerequistes
To get the most out of the presented material, I suggest a strong programming background in at least one of the following (Python, R, Julia, or Matlab), experience with linear algebra and basic probability and statistics. Do not worry if you don't know much biology. We will have a gentle introduction at the beginning and throughout. Please feel free to talk to me about any of these. 

## Course Structure
This course will be mostly lecture-based with two homework assignments and a course project. I will provide ideas for several publicly available biological datasets and open problems for you to work on. This will give you an opportunity to implement/apply methodology discussed in the papers that we will discuss together. You are welcome to work on teams for this project.

Most of the lectures will be based around several papers. To benefit your own understanding, I will provide a set of questions that should be answered for one of the papers discussed in each lecture.

## Homework
There will be two homework assignments to translate concepts into code. Often, things can become a bit easier to understand and use when they are implemented by you. :) I will be happy to read/run code written in Python, R, Julia, or Matlab. Please submit your homework writeup as a PDF. 

## Background Resources
Most of what we discuss in class will come from papers. However, I suggest the following textbooks as background references. Conveniently, they are also available for free.

* Pattern Recognition and Machine Learning-- Chris Bishop // http://users.isr.ist.utl.pt/~wurmd/Livros/school/Bishop%20-%20Pattern%20Recognition%20And%20Machine%20Learning%20-%20Springer%20%202006.pdf

* All of Statistics-- Larry Wasserman // http://www.stat.cmu.edu/~larry/all-of-statistics/

* The Matrix Cookbook // https://www.math.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf

## Readings
For each class, I will share the papers that we will go over. You will only be required to write a summary of one of the potentially multiple papers assigned for that day.

## Final Project
I will provide you with several examples of publicly available biological datasets and problems. Half-way through the semester, you will submit your project proposal and present your idea to the class.  The proposal will be a short document describing 1) The problem 2) A background on other people's attempts to solve this problem and 3) A background on your idea of a solution and 4) Data you will use to test your method. At the end of the semester you will write a short paper explaining your method and results and present your results.

# Schedule

| Date | Topic | Reading |
|------|-------|---------|
| Week 1-1    | Intro, biological primer, bioinformatics vs comp bio, big picture challenges     |        |
| Week 1-2    | Review of Basic ML concepts (cross-validation, classification, regression, clustering)     |        |
| Week 2-2    | Building graphs from data, matrix factorization | |
| Week 3-1 | Partitioning Graphs, Ranking Nodes | |
| Week 4-1 | Graph Embeddings: Embedding Nodes and Embedding Edges | Node2Vec and Edge2Vec |
| Week 4-2 | Single Cell Day 1: Intro to single-cell proteomics, visualization, normalization and batch effect correction | spade |
|Week 5-1 | Single Cell Day 2: Graph-based analysis of single-cell data | phenograph |
| Week 5-2 | Single Cell Day 3: Feature Engineering from single-cell data and linking to external variables | Citrus, VoPo | 
| Week 6-1 | Single Cell Day 4: Differential Analysis of Cell-Populations | Diffcyt, Cydar |
| Week 6-2 | Single Cell Day 5: Graph-based matching of single-cell data | Conos, LIGER
| Week 7-1 | Single Cell Day 6: Deep Learning in Single-Cell Analysis | SAUCIE, PHATE, CellCNN |
| Week 7-2 | Single Cell Day 7: Trajectory Inference | |
| Week 8-1 | Single Cell Day 8: Benchmarking in Trajectory Inference | |
| Week 8-2 : Project Proposals Due | Presentations of Project Propsals Day 1 | |
| Week 9-1 | Project Proposal Presentation Day 2 | |
| Week 9-2| Single Cell Day 9: Benchmarking in Single-Cell Analysis | Aghaeepour et al |
| Week 10-1 | Single Cell Day 10: Imaging Proteomics + Spatial Regularization : computational challeneges in combining tissue images and protein expression | |
| Week 10-2 | Multiomics Day 1: Constructing a joint embedding of samples according to multiple modalities | SNF, grassmann embed |  
| Week 11-1 | Multiomics Day 2: MOFA-1 and MOFA-2: Multiomics Factor Analysis | MOFA-1, MOFA-2 | 
| Week 11-2 | Multiomics Day 3: Uncovering Relationships Between Modalities | mmvec| 
| Week 12-1 | Multiomics Day 4: Stacked Generalization and CCA in multiomics studies | Ghaemi |
| Week 12-2 | Multiomics Day 5: Benchmarking in multiomics studies | |
|Week 13-1  | Incorporating Prior Biological Knowledge into Analysis | |
|Week 13-2 |  Systems Immunology Topic: TCR/BCR (T and B cell reperotire analysis) | | 
|Week 14-1 | Partial Correlation, Thresholding etc for Identifying Meaningful Interactions | | 
|Week 14-2 | Enrichment Analysis, writing for an interdiscplinary audience | | 
|Week 15-1 | Project Presentations Day 1 | |
|Week 15-2 | Project Presentations Day 2 | |
|Final Exam Day | Project papers due | |

